Flatpak

mattdangerw opened the pull request:
https://phabricator.endlessm.com/T13554

mattdangerw committed:
Initial files

Adapted from
https://git.gnome.org/browse/gnome-sdk-images/

mattdangerw committed:
runtime: add modules

The udev module is grabbed from
https://github.com/alexlarsson/spotify-app/blob/master/com.spotify.Client.json

udev isn't stable so bringing it into a runtime is questionable.
But chromium needs it, we need to support the electron prebuilt
binaries with chromium, so I don't see an alternative.

We are building the latest stable node only for the Sdk. Since electron
uses its own copy of v8 to run its main process, we shouldn't need
node in the runtime.

mattdangerw committed:
runtime: add libappindicator dependency

Adapted from
https://github.com/alexlarsson/nightly-build-apps/blob/master/org.telegram.TelegramDesktopDevel.json
with an extra patch as we don't have gtk-doc-tools available.

mattdangerw commented on metadata.sdk.in:
```diff
@@ -0,0 +1,22 @@
+version=@@SDK_RUNTIME_VERSION@@
+directory=share/zoneinfo
+
+[Extension org.freedesktop.Platform.GStreamer]
```
Should I remove this? Copied it in from the gnome runtime, but I don't think chromium uses gstreamer...

cosimoc replied:
I don't think it hurts to keep it here.

cosimoc commented on io.atom.electron.Sdk.json.in:
```diff
@@ -0,0 +1,268 @@
+            ]
+        },
+        {
+            "name": "libpgp-error",
```
Should be `libgpg-error`

cosimoc commented on os-release.in:
```diff
@@ -0,0 +1,4 @@
+NAME=ELECTRON
```
Change to not all caps

cosimoc commented on io.atom.electron.Sdk.json.in:
```diff
@@ -0,0 +1,268 @@
+                },
+                {
+                    "type": "patch",
+                    "path": "libdbusmenu-remove-gtk-doc-init.patch"
```
You can add this to avoid this patch: https://git.gnome.org/browse/gnome-doc-utils/
But if you don't, the patch should be renamed to mention `gnome-doc` instead of `gtk-doc`

cosimoc replied:
You can use this instead: https://launchpad.net/libdbusmenu/16.04/16.04.0/+download/libdbusmenu-16.04.0.tar.gz

cosimoc commented on io.atom.electron.Sdk.json.in:
```diff
@@ -0,0 +1,268 @@
+                    "sha256": "0d15cec3b6d55c60eac205b1f3ba81a1ed4eadd9d0f8e7c508bc7065d0c4ca50"
+                }
+            ]
+        },
```
Also need `gnome-themes-standard` like this, because that will get you the GTK2 theme.
https://git.gnome.org/browse/gnome-sdk-images/tree/org.gnome.Sdk.json.in#n556

ptomato commented on os-release-configure:
```diff
@@ -0,0 +1,9 @@
+
+install:
+	mkdir -p ${DESTDIR}/usr/lib
+		install os-release ${DESTDIR}/usr/lib
```
two tabs?

cosimoc commented on io.atom.electron.Sdk.json.in:
```diff
@@ -0,0 +1,275 @@
+                "env": {
+                    "HAVE_VALGRIND_TRUE": "'#'",
+                    "HAVE_VALGRIND_FALSE": "''"
+                }
```
Are these still needed?

ptomato commented:
Maybe consider adding a `flatpak install` of `org.freedesktop.Sdk 1.4` and `org.freedesktop.Sdk.Debug 1.4` to the makefile, or to a separate target so you can do `make install-runtimes`?

mattdangerw committed:
runtime: Bump to latest versions of libnotify and drop the patch

mattdangerw committed:
Build a "base app" instead of runtime

We will build a "base app" instead of a runtime for electron app
dependencies. Essentially moving everything specified here from
/usr to /app and distributing them with the application. On its
own this base app will be pretty useless, but it can be specified
as a base for another flatpak build, a feature that just landed
in flatpak-builder.

This should be easier to maintain than an actual runtime. i.e.
freedesktop can push our a critical security fix for the platform
without us needing to rebuild this base app.

Relevant discussion here
https://github.com/flatpak/flatpak/issues/347

mattdangerw committed:
BaseApp: remove node

Node is useful as a build tool, but since electron comes with its
own node binaries, is not useful to stick in an actual app install.

Eventually would be nice to find an easy way to expose a build tool
like this to the electron app developer, but it does not belong in
the base app

mattdangerw committed:
BaseApp: remove os-release file

This only makes sense for runtimes

mattdangerw committed:
BaseApp: remove gnome-common

We don't need it anymore

mattdangerw committed:
BaseApp: add xdg-open shim

mattdangerw commented:
> Maybe consider adding a flatpak install of org.freedesktop.Sdk 1.4 and org.freedesktop.Sdk.Debug 1.4 to the makefile, or to a separate target so you can do make install-runtimes?

Not sure of a great way to do this, you'd have to assume a name for the gnome repo (`gnome`, `gnome-nighty`?). I think for now, best to have people setup runtimes themselves. I can add a README

Am building fancier tools for auto download update of runtimes into the electron app tooling, but not sure we need that here as well.

mattdangerw commented:
Actually, probably a bit easier now with new `--from` flatpak feature and https://sdk.gnome.org/gnome.flatpakrepo. I'll look at adding an optional make rule for installing deps.

mattdangerw committed:
Makefile: helper to install deps for you

Won't be useful if you already have a gnome remote or freedesktop
runtimes, but helpful on a fresh system

ptomato commented on Makefile:
```diff
@@ -0,0 +1,16 @@
+ARCH ?= $(shell flatpak --default-arch)
+REPO ?= repo
+
+# Convenience to install deps. If you already have a gnome just install from that!
```
Might be good to note in the comment here that this requires flatpak 0.6.13

ptomato commented on Makefile:
```diff
@@ -0,0 +1,16 @@
+	flatpak --user remote-add --from gnome https://sdk.gnome.org/gnome.flatpakrepo
+	flatpak --user install gnome org.freedesktop.Platform//1.4 org.freedesktop.Sdk//1.4
+
+all: ${REPO}
```
`all` should be the first rule; otherwise `make` with no target will pick `install-deps`

ptomato reviewed (commented):
Except for the install-deps rule I built this successfully with Flatpak 0.6.11.

cosimoc reviewed (commented):
No other comments from my side -- looks good! :+1:

mattdangerw committed:
README: fill in with some actual documentation

alexlarsson commented on io.atom.electron.BaseApp.json:
```diff
@@ -0,0 +1,229 @@
+                    "sha256": "9268e1cc487de5e6e4460fca612a06e4f383072ac43ae90603e5e46783d3e540"
+                }
+            ]
+        },
```
You should not need libgpg-error as it is in the runtime.

alexlarsson commented on io.atom.electron.BaseApp.json:
```diff
@@ -0,0 +1,229 @@
+                    "sha256": "ddac6111077d0a1612247587be238c5294dd0ee4d76dc7ba783cc55fb0337071"
+                }
+            ]
+        },
```
Same with libgcrypt. However, the yocto base is a bit weird in that it breaks the libgcrypt-config script:

```
$ /usr/bin/libgcrypt-config 
ERROR: /usr/bin/libgcrypt-config should not be used, use an alternative such as pkg-config
--should-not-have-used-/usr/bin/libgcrypt-config
```

So, you might have to patch the build to use pkg-config to find it.

mattdangerw committed:
BaseApp: don't build gcrypt

Its already built into freedesktop. We do need to rework
libgnome-keyring to find the installed library with pkg-config though.
